### PR TITLE
feat(kafka acl): add convenience commands

### PR DESCRIPTION
**Pull request by rkpattnaik780**
### Verification Steps

1. `rhoas kafka acl grant-permissions --producer --user user_name --topic "*"` should create ACL rules allowing user to write to all topics in the instance.
2. `rhoas kafka acl grant-permissions --consumer --user user_name --topic-prefix "abc"` should create ACL rules allowing user to consume from all topics starting with "abc".

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation change
- [ ] Other (please specify)

### Checklist

- [X] Documentation added for the feature
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer

### CODE REVIEW

File: pkg/cmd/kafka/acl/consumer/consumer.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+			if opts.user != "" && opts.svcAccount != "" {
+				return opts.localizer.MustLocalizeError("kafka.acl.common.error.bothPricipalsSelected")
+			}
+
+			return runConsumer(opts)
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.user, "user", "", opts.localizer.MustLocalize("kafka.acl.common.flag.user.description"))
+	cmd.Flags().StringVar(&opts.user, "service-account", "", opts.localizer.MustLocalize("kafka.acl.common.flag.serviceAccount.description"))
```
Comment (craicoverflow) [inline pkg/cmd/kafka/acl/consumer/consumer.go:77]: I know this is what is described in the doc, but I do fear that is brings complexity by having too many options.

User:

```shell
rhoas kafka acl consumer --principal User::ephelan_kafka_devexp
```

ServiceAccount:

```shell
rhoas kafka acl consumer --principal ServiceAccount::ephelan_kafka_devexp
```

Maybe we can discuss this on the call today.

File: pkg/cmd/kafka/acl/consumer/consumer.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+	cmd.Flags().StringVar(&opts.topic, "topic-prefix", "", opts.localizer.MustLocalize("kafka.acl.common.flag.topicPrefix.description"))
```
[the remark this answers is not in the document]
Comment (rkpattnaik780) [reply pkg/cmd/kafka/acl/consumer/consumer.go:78]: have added this to agenda, we might need different flags for each resource type

**Comment by wtrocki** [comment]
Separate PR:
[kafka.acl.list.cmd.longDescription] - let's copy what we have in the documentation to make this more documented. 
We can forget about this later

File: pkg/cmd/kafka/acl/consumer/consumer.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+		case http.StatusInternalServerError:
+			return opts.localizer.MustLocalizeError("kafka.acl.common.error.internalServerError")
+		case http.StatusServiceUnavailable:
+			return opts.localizer.MustLocalizeError("kafka.acl.common.error.unableToConnectToKafka", localize.NewEntry("Name", kafkaInstance.GetName()))
+		default:
+			return err
+		}
+	}
+
+	aclBindGroupRead := *kafkainstanceclient.NewAclBinding(kafkainstanceclient.ACLRESOURCETYPE_GROUP, "*", patternArg, userArg, kafkainstanceclient.ACLOPERATION_READ, kafkainstanceclient.ACLPERMISSIONTYPE_ALLOW)
```
Comment (wtrocki) [inline pkg/cmd/kafka/acl/consumer/consumer.go:177]: binding in name gives me flashbacks :D

File: pkg/localize/locales/en/cmd/acl.en.toml [hunk 4]
```diff
@@ -29,6 +29,24 @@ one = 'unable to connect to Kafka instance "{{.Name}}"'
...
 
+[kafka.acl.common.error.bothPricipalsSelected]
+one = 'flags "user" and "service-account" can not be provided simultaneously'
+
+[kafka.acl.common.error.noPricipalsSelected]
+one = 'principal is missing, provide "user" or "service-account" flag'
+
+
+[kafka.acl.common.flag.user.description]
+one = 'username to be used as principal'
```
Comment (bhardesty) [inline pkg/localize/locales/en/cmd/acl.en.toml:40]:
```suggestion
one = 'Username to be used as principal'
```

File: pkg/localize/locales/en/cmd/acl.en.toml [hunk 5]
```diff
@@ -29,6 +29,24 @@ one = 'unable to connect to Kafka instance "{{.Name}}"'
...
+
+[kafka.acl.common.flag.serviceAccount.description]
+one = 'service account ID to be used as principal'
```
Comment (bhardesty) [inline pkg/localize/locales/en/cmd/acl.en.toml:43]:
```suggestion
one = 'Service account ID to be used as principal'
```

File: pkg/localize/locales/en/cmd/acl.en.toml [hunk 6]
```diff
@@ -29,6 +29,24 @@ one = 'unable to connect to Kafka instance "{{.Name}}"'
...
+
+[kafka.acl.common.flag.topicPrefix.description]
+one = 'prefix topic name to define ACL rules for'
```
Comment (bhardesty) [inline pkg/localize/locales/en/cmd/acl.en.toml:46]:
```suggestion
one = 'The prefix topic name to define ACL rules for'
```

File: pkg/localize/locales/en/cmd/acl.en.toml [hunk 7]
```diff
@@ -29,6 +29,24 @@ one = 'unable to connect to Kafka instance "{{.Name}}"'
...
+
+[kafka.acl.common.flag.transactionalID.description]
+one = 'transactional ID for which ACL rules are to be defined'
```
Comment (bhardesty) [inline pkg/localize/locales/en/cmd/acl.en.toml:49]:
```suggestion
one = Transactional ID for which ACL rules are to be defined'
```

File: pkg/localize/locales/en/cmd/acl.en.toml [hunk 8]
```diff
@@ -64,4 +82,41 @@ one = 'All accounts'
...
-one = 'starts with'
\ No newline at end of file
+one = 'starts with'
+
+
+[kafka.acl.producer.cmd.shortDescription]
+one = 'Add ACL rules for user to produce messages to topics'
+
+[kafka.acl.producer.cmd.longDescription]
+one = 'This command will create ACL rules to allow user to produce messages to topics'
```
Comment (bhardesty) [inline pkg/localize/locales/en/cmd/acl.en.toml:92]:
```suggestion
one = 'This command will create ACL rules to allow a particular user to produce messages to topics'
```

File: pkg/localize/locales/en/cmd/acl.en.toml [hunk 9]
```diff
@@ -64,4 +82,41 @@ one = 'All accounts'
...
+
+[kafka.acl.producer.cmd.example]
+one = '''
+# Grant access to principal to allow producing messages to all topics
```
Comment (bhardesty) [inline pkg/localize/locales/en/cmd/acl.en.toml:96]:
```suggestion
# Grant access to principal for producing messages to all topics
```

File: pkg/localize/locales/en/cmd/acl.en.toml [hunk 10]
```diff
@@ -64,4 +82,41 @@ one = 'All accounts'
...
+$ rhoas kafka acl producer --user user_name
+
+# Grant access to principal to allow producing messages to topics starting with "abc"
```
Comment (bhardesty) [inline pkg/localize/locales/en/cmd/acl.en.toml:99]:
```suggestion
# Grant access to principal for producing messages to topics starting with "abc"
```

File: pkg/localize/locales/en/cmd/acl.en.toml [hunk 11]
```diff
@@ -64,4 +82,41 @@ one = 'All accounts'
 one = 'is'
 
 [kafka.acl.list.startsWith]
-one = 'starts with'
\ No newline at end of file
+one = 'starts with'
+
+
+[kafka.acl.producer.cmd.shortDescription]
+one = 'Add ACL rules for user to produce messages to topics'
```
Comment (bhardesty) [inline pkg/localize/locales/en/cmd/acl.en.toml:89]:
```suggestion
one = 'Add ACL rules to grant users access to produce messages to topics'
```

File: pkg/localize/locales/en/cmd/acl.en.toml [hunk 12]
```diff
@@ -64,4 +82,41 @@ one = 'All accounts'
...
+one = '''
+# Grant access to principal to allow producing messages to all topics
+$ rhoas kafka acl producer --user user_name
+
+# Grant access to principal to allow producing messages to topics starting with "abc"
+$ rhoas kafka acl producer --user user_name --topic-prefix "abc"
+'''
+
+[kafka.acl.producer.log.info.aclsCreated]
+one = 'ACLs created to allow principal to write messages to topics'
```
Comment (bhardesty) [inline pkg/localize/locales/en/cmd/acl.en.toml:104]:
```suggestion
one = 'ACLs created to enable principal to write messages to topics'
```

File: pkg/localize/locales/en/cmd/acl.en.toml [hunk 13]
```diff
@@ -64,4 +82,41 @@ one = 'All accounts'
...
+
+[kafka.acl.consumer.cmd.shortDescription]
+one = 'Add ACL rules for user to consume messages from topics'
```
Comment (bhardesty) [inline pkg/localize/locales/en/cmd/acl.en.toml:107]:
```suggestion
one = 'Add ACL rules to grant users access to consume messages from topics'
```

File: pkg/localize/locales/en/cmd/acl.en.toml [hunk 14]
```diff
@@ -64,4 +82,41 @@ one = 'All accounts'
...
+
+[kafka.acl.consumer.cmd.longDescription]
+one = 'This command will create ACL rules to allow user to consume messages from topics'
```
Comment (bhardesty) [inline pkg/localize/locales/en/cmd/acl.en.toml:110]:
```suggestion
one = 'This command creates ACL rules that grant the specified user access to consume messages from topics'
```

File: pkg/localize/locales/en/cmd/acl.en.toml [hunk 15]
```diff
@@ -64,4 +82,41 @@ one = 'All accounts'
...
+
+[kafka.acl.consumer.cmd.example]
+one = '''
+# Grant access to principal to allow consuming messages from all topics
```
Comment (bhardesty) [inline pkg/localize/locales/en/cmd/acl.en.toml:114]:
```suggestion
# Grant access to principal for consuming messages from all topics
```

File: pkg/localize/locales/en/cmd/acl.en.toml [hunk 16]
```diff
@@ -64,4 +82,41 @@ one = 'All accounts'
...
+$ rhoas kafka acl consumer --user user_name
+
+# Grant access to principal to consume messages from topics starting with "abc"
```
Comment (bhardesty) [inline pkg/localize/locales/en/cmd/acl.en.toml:117]:
```suggestion
# Grant access to principal for consuming messages from topics starting with "abc"
```

File: pkg/localize/locales/en/cmd/acl.en.toml [hunk 17]
```diff
@@ -64,4 +82,41 @@ one = 'All accounts'
...
+$ rhoas kafka acl consumer --user user_name --topic-prefix "abc"
+'''
+
+[kafka.acl.consumer.log.info.aclsCreated]
+one = 'ACLs created to allow principal to consume messages from topics'
```
Comment (bhardesty) [inline pkg/localize/locales/en/cmd/acl.en.toml:122]:
```suggestion
one = 'ACLs created to enable principal to consume messages from topics'
```

File: pkg/cmd/kafka/acl/grant/grant.go [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+	}
+
+	if opts.svcAccount != "" {
+		userArg = buildPrincipal(opts.svcAccount)
+	}
+
+	req := api.AclsApi.CreateAcl(opts.Context)
+
+	if opts.producer || opts.consumer {
+		aclBindTopicDescribe := *kafkainstanceclient.NewAclBinding(kafkainstanceclient.ACLRESOURCETYPE_TOPIC, topicNameArg, patternArg, userArg, kafkainstanceclient.ACLOPERATION_DESCRIBE, kafkainstanceclient.ACLPERMISSIONTYPE_ALLOW)
```
Comment (wtrocki) [inline pkg/cmd/kafka/acl/grant/grant.go:133]:
```suggestion
		aclBindTopicDescribe := *kafkainstanceclient.NewAclBinding(kafkainstanceclient.ACLRESOURCETYPE_TOPIC, 
		topicNameArg, patternArg, userArg, kafkainstanceclient.ACLOPERATION_DESCRIBE, kafkainstanceclient.ACLPERMISSIONTYPE_ALLOW)
```

File: pkg/cmd/kafka/acl/grant/grant.go [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+
+		req = req.AclBinding(aclBindTopicDescribe)
+
+		err = acl.ExecuteACLRuleCreate(req, opts.localizer, kafkaName)
+		if err != nil {
+			return err
```
Comment (wtrocki) [inline pkg/cmd/kafka/acl/grant/grant.go:139]: That needs mapping to something more user-friendly. Backend errors by default will only include status code (no actual info what went wrong)

**Comment by wtrocki** [comment]
All ACL rules are aligned with the Kafka specification/patterns for consuming and producing content. 
Great job! Little bit refinement and we can get that in.

For redability I would suggest to define ACL in the group and then have single method that tries to execute them in the loop. That will allow us to swap easily to batch API when available with single like of code.

File: pkg/cmd/kafka/acl/grant/grant.go [hunk 20]
```diff
@@ -0,0 +1,228 @@
...
+			if opts.user != "" && opts.svcAccount != "" {
+				return opts.localizer.MustLocalizeError("kafka.acl.common.error.bothPricipalsSelected")
+			}
+
+			return runGrantPermissions(opts)
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.user, "user", "", opts.localizer.MustLocalize("kafka.acl.grantPermissions.flag.user.description"))
+	cmd.Flags().StringVar(&opts.svcAccount, "service-account", "", opts.localizer.MustLocalize("kafka.acl.grantPermissions.flag.serviceAccount.description"))
```
Comment (craicoverflow) [inline pkg/cmd/kafka/acl/grant/grant.go:82]: We need to be clear that the service account client ID is used for this, not the service account ID (confusing as hell having two).

Could we merge these into a single flag like the UI does?

Comment (wtrocki) [reply pkg/cmd/kafka/acl/grant/grant.go:82]: btw.. validation is not required immediately/right now and can be added later phase over released commands

File: pkg/cmd/kafka/acl/grant/grant.go [hunk 21]
```diff
@@ -74,17 +74,29 @@ func NewGrantPermissionsACLCommand(f *factory.Factory) *cobra.Command {
...
 
+			if opts.admin && opts.user == "" {
+				return opts.localizer.MustLocalizeError("kafka.acl.grantPermissions.error.user.required")
+			}
+
+			if opts.admin && (opts.consumer || opts.producer) {
+				return opts.localizer.MustLocalizeError("kafka.acl.grantPermissions.admin.error.notAllowed")
+			}
+
+			if !opts.consumer && opts.group != "" {
```
Comment (wtrocki) [inline pkg/cmd/kafka/acl/grant/grant.go:85]: We need case when consumer is true but group missing.

**Comment by craicoverflow** [comment]
```shell
$ rhoas kafka acl grant-permissions --service-account srvc1 --consumer --topic-prefix "*" --group-prefix "*"
❌ Undefined response type. Run the command in verbose mode using the -v flag to see more information
```

I'm seeing this error again, perhaps @wtrocki's changes were lost from the most recent commit which @rkpattnaik780 did not have locally?

**Pull request merged**